### DISCUSSION TITLE
Different directory structure

### DISCUSSION THREAD

**Discussion by fedemengo-genuino** [adonisjs/core#2462] [Help]
I was wondering if it's possible to setup an adonis project to use a different directory structure, specifically I'd prefer to have my files in the following way
```
.
├── src/
  ├── User/
    ├── Controllers
    ├── Exception
    ├── Middleware
    ├── Models
    └── Validators
  ├── Post/
    ├── Controllers
    ├── Exception
    ├── Middleware
    ├── Models
    └── Validators
├── start/
  ├── app.js
  ├── kernel.js
  └── routes.js
├── test/
├── ace
├── server.js
└── package.json
```
So basically I'd like to have controllers/validators/middleware/etc grouped by the entity they are managing

## Package version
Lucid 6.2.1 with Adonis 4.1.0

## Node.js and npm version
v15.11.0 and 7.6.3

**Comment by RomainLanz** [comment]
Hey @fedemengo-genuino! 👋 

Yes! You can move your `Controllers`, `Exceptions`, `Listeners`, `Models`, etc. to any folders your want!

However, since you are not using the default architecture, you will need to specify where to find your files.

For example, you will need to specify the namespace of your controller instead of only typing their name.

```ts
Route.group(() => {
  // Will load `App/User/Controller/UsersController.ts`
  Route.get('/', 'UsersController.index')
}).namespace('App/User/Controllers')
```

📚 [Route's Controller Namespace](https://preview.adonisjs.com/guides/http/routing#routes-controller-namespace)

**Reply by Agilulfo1820** [reply]
Hey @RomainLanz what if I would like to do the same with the routes? Instead of having all my routes in the route file, I would like to have a specific file for the admin routes, one for the member routes, one for guests, and so on. Should be simple to do but had some problem on figuring out how to do it.

**Reply by RomainLanz** [reply]
AdonisJS is only JavaScript/TypeScript after all!

So create your file anywhere you want, and load them inside your main route file.

Let say you have those files:
```
start/routes/admin.ts
start/routes/members.ts
start/routes/guest.ts
```

Then you can import them in your main file:

```ts
// start/routes.ts
import './routes/admin'
import './routes/members'
import './routes/guest'
```

Also, if you prefer to not have a main file, you can directly add your routes file inside the `preload` key of your `.adonisjsrc.json` file.

**Comment by fedemengo-genuino** [comment]
Is there anything special about `App`? Because I wanted to move out the `app` directory altogether, I created a new directory `src` at the same level of `app` 

```
.
├── app/
├── src/
  ├── User/
    ├── Controllers
        └── UserController.js
    ├── Exception
    ├── Middleware
        └── SignUpMiddleware
    ├── Models
    ├── Validators
        └── SignUpValidator
├── start/
  ├── app.js
  ├── kernel.js
  └── routes.js
├── test/
├── ace
├── server.js
└── package.json
```
And I created a controller `src/User/Controllers/UserController.js` then I defined a new route as such

```
Route.group('User group', () => {
    Route.post('/user', 'Controllers/UserController.store')
}).prefix('/api/user/v1')
  .namespace('src/User/')
```
My idea is to use something like

```
Route.group('User group', () => {
    Route.post('/user', 'Controllers/UserController.store')
        .middleware('Middleware/SignUpMiddleware')    
        .validator('Validator/SignUpValidator')
}).prefix('/api/user/v1')
  .namespace('src/User/')
```
But just by putting the controller I get `Cannot find module '/srv/app/app/Controllers/Http/src/User/Controllers/UserController' `

I also tried adding 

```
  "autoload": {
    "App": "./app",
    "Src": "./src"
  }
```
In `package.json`

**Reply by Agilulfo1820** [reply]
@RomainLanz Could you please take a look at this? Thanks!

**Reply by thetutlage** [reply]
Run `node ace dump:rcfile` and you can see the default list of directories and namespaces and just overwrite the once you want

**Reply by fedemengo-genuino** [reply]
@thetutlage Thanks for the help, but it says `error: 'dump:rcfile' is not a registered command`, I just want to point out the project is based on **v4.1**

**Reply by Agilulfo1820** [reply]
@thetutlage sorry for bugging you, but could you please assist us with this issue?

**Reply by thetutlage** [reply]
As stated the app is on 4.1, there are certain restrictions in place related to ace commands mainly.

You can create your controllers, models, event listeners anywhere in the app you want and it should work fine. However, the ace command have hardcoded paths and it cannot be changed.

The `.adonisrc.json` customizations works with v5 only.

---

Feel free to ask if any specific thing doesn't work

**Reply by fedemengo-genuino** [reply]
@thetutlage thanks for the answer, a specific example of something I tried to do is what I posted  [above](https://github.com/adonisjs/core/discussions/2462#discussioncomment-585142)

**Reply by thetutlage** [reply]
That shares your expected outcome for the entire project structure.

What I meant, if you face issues defining controllers in routes or models, then you should initiate a conversation for that particular thing

**Comment by dqgiang** [comment]
In my projects (both v4.1 and v5), I have similar setup with routes, controllers, validators, models, listeners, etc. grouped into modules. The loading of all those stuff can be done inside each module using the Service Provider. Check [v4 docs](https://adonisjs.com/docs/4.1/service-providers) (afaik, docs for this topic in v5 is missing but you can check example from lucid package).

With this approach, all you have to do to wire them to the app is declaring the provider in `.adonisrc.json` (or `app.js` if using v4.1):

```jsonc
"providers": [
    "./providers/AppProvider",
    "@adonisjs/core",
    "@adonisjs/lucid",
    "./modules/MyModule/providers/MyModuleProvider", //👈 Here 
  ],
```

Provider example (v4.1 will be similar):
```ts
// modules/MyModule/providers/MyModuleProvider.ts

import { ApplicationContract } from '@ioc:Adonis/Core/Application';

export default class MyModuleProvider {
  public static needsApplication = true;
  constructor(protected _app: ApplicationContract) {}

  private async _loadConfig() { /* Load from `config` directory */ } 
  private async _registerBindings() { /* Register ioc bindings */}

  public async register() {
    await this._loadConfig();
    await this._registerBindings();
  }

  private async _registerEvents() {
    await import('../start/events'); // Your module event subscription
  }

  private async _registerRoutes() {
    await import('../start/routes'); // Your module routes
  }

  public async _registerRules() {
    await import('../start/module-custom-validation-rules');
  }

  public async boot() {
    await this._registerEvents();
    await this._registerRules();
    await this._registerRoutes();
  }
// ...
}
```

You can even make custom commands, extending from the base `make:*` commands to generate files directly to folders in your module.

```
node ace mymake:controller ControllerName ModuleName
```

**Reply by fedemengo-genuino** [reply]
Thanks for the tips, I'll give it a shot

**Comment by junkeritechnepal** [comment]
adonis is great, except with this folder structuring and importing my classes ( controllers, services... import('#controllers/...) is error prone )

**Comment by Akeem92** [comment]
Heyy @RomainLanz. Please just for know do you using any king of api docs with adonisJs ?
Am trying to use https://github.com/ad-on-is/adonis-autoswagger but i have Modular architecture so i can't remap where doc find my validators files.
